<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Test--requests.post-on-Michigan's-site" data-toc-modified-id="Test--requests.post-on-Michigan's-site-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Test  requests.post on Michigan's site</a></span><ul class="toc-item"><li><span><a href="#First-test-county-look-up" data-toc-modified-id="First-test-county-look-up-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>First test county look-up</a></span></li><li><span><a href="#Then-test-township-look-up" data-toc-modified-id="Then-test-township-look-up-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Then test township look-up</a></span></li></ul></li><li><span><a href="#Scrape-all-county-pages" data-toc-modified-id="Scrape-all-county-pages-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Scrape all county pages</a></span></li><li><span><a href="#Analyze-the-county-pages" data-toc-modified-id="Analyze-the-county-pages-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Analyze the county pages</a></span></li><li><span><a href="#Build-a-list-of-all-county-names" data-toc-modified-id="Build-a-list-of-all-county-names-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Build a list of all county names</a></span></li><li><span><a href="#Build-a-list-of-all-jurisdictionCodes" data-toc-modified-id="Build-a-list-of-all-jurisdictionCodes-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Build a list of all jurisdictionCodes</a></span></li><li><span><a href="#Build-a-list-of-all-township-names" data-toc-modified-id="Build-a-list-of-all-township-names-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Build a list of all township names</a></span></li><li><span><a href="#Scrape-all-township-(jurisdictions)-pages" data-toc-modified-id="Scrape-all-township-(jurisdictions)-pages-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Scrape all township (jurisdictions) pages</a></span></li><li><span><a href="#Analyze-the-township-pages" data-toc-modified-id="Analyze-the-township-pages-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Analyze the township pages</a></span><ul class="toc-item"><li><span><a href="#Isolate-the-textual-data-from-the-maps" data-toc-modified-id="Isolate-the-textual-data-from-the-maps-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>Isolate the textual data from the maps</a></span></li></ul></li><li><span><a href="#Disagregate-address-elements-into-separate-records" data-toc-modified-id="Disagregate-address-elements-into-separate-records-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Disagregate address elements into separate records</a></span></li><li><span><a href="#Create-a-DataFrame-to-hold-township-address-data" data-toc-modified-id="Create-a-DataFrame-to-hold-township-address-data-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Create a DataFrame to hold township address data</a></span></li><li><span><a href="#Write-township-data-to-csv-file" data-toc-modified-id="Write-township-data-to-csv-file-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>Write township data to csv file</a></span></li><li><span><a href="#Disagregate-dropbox-address-data" data-toc-modified-id="Disagregate-dropbox-address-data-12"><span class="toc-item-num">12&nbsp;&nbsp;</span>Disagregate dropbox address data</a></span></li><li><span><a href="#Determine-patterns-of-data-in-dropbox-rows" data-toc-modified-id="Determine-patterns-of-data-in-dropbox-rows-13"><span class="toc-item-num">13&nbsp;&nbsp;</span>Determine patterns of data in dropbox rows</a></span><ul class="toc-item"><li><span><a href="#Analyze-first-set" data-toc-modified-id="Analyze-first-set-13.1"><span class="toc-item-num">13.1&nbsp;&nbsp;</span>Analyze first set</a></span></li><li><span><a href="#Analyze-second-set" data-toc-modified-id="Analyze-second-set-13.2"><span class="toc-item-num">13.2&nbsp;&nbsp;</span>Analyze second set</a></span></li></ul></li><li><span><a href="#Loop-thru-all-cases-to-get-dropbox-data" data-toc-modified-id="Loop-thru-all-cases-to-get-dropbox-data-14"><span class="toc-item-num">14&nbsp;&nbsp;</span>Loop thru all cases to get dropbox data</a></span></li><li><span><a href="#Create-DataFrame-to-hold-drop-box-address-data" data-toc-modified-id="Create-DataFrame-to-hold-drop-box-address-data-15"><span class="toc-item-num">15&nbsp;&nbsp;</span>Create DataFrame to hold drop box address data</a></span></li><li><span><a href="#Create-dropbox-Address-and-City" data-toc-modified-id="Create-dropbox-Address-and-City-16"><span class="toc-item-num">16&nbsp;&nbsp;</span>Create dropbox Address and City</a></span></li><li><span><a href="#Create-dropbox-location_type" data-toc-modified-id="Create-dropbox-location_type-17"><span class="toc-item-num">17&nbsp;&nbsp;</span>Create dropbox location_type</a></span></li><li><span><a href="#Create-dropbox-notes" data-toc-modified-id="Create-dropbox-notes-18"><span class="toc-item-num">18&nbsp;&nbsp;</span>Create dropbox notes</a></span></li><li><span><a href="#Write-dropbox-file-to-csv-file" data-toc-modified-id="Write-dropbox-file-to-csv-file-19"><span class="toc-item-num">19&nbsp;&nbsp;</span>Write dropbox file to csv file</a></span></li></ul></div>

In [1]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
import re
import json

# Test  requests.post on Michigan's site

## First test county look-up

In [2]:
payload = {'CountyID':'1'}
res = requests.post(f'https://mvic.sos.state.mi.us/Voter/SearchByCounty', data=payload, verify=False)

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,


In [3]:
res.status_code

200

In [4]:
type(res)

requests.models.Response

In [5]:
soup = BeautifulSoup(res.content, 'html.parser')

In [6]:
soup


<input id="hfPageTitle" name="hfPageTitle" type="hidden" value="County Clerk"/><script async="" defer="" src="https://www.bing.com/api/maps/mapcontrol?callback=GetMap" type="text/javascript"></script>
<script src="/bundles/BingMapJS?v=XgqZHpFMh6PgxqFv2EbFUE-JrwdZK88VAIk5Ee20hLA1"></script>
<input id="hfPageTitle" name="hfPageTitle" type="hidden" value="Your city/township clerk"/>
<input id="hfPageName" name="hfPageName" type="hidden" value="clerk"/>
<input id="hfClerkLatitude" name="hfClerkLatitude" type="hidden" value="44.657546"/>
<input id="hfClerkLongitude" name="hfClerkLongitude" type="hidden" value="-83.296131"/>
<input id="hfClerkRouteRequestTitle" name="hfClerkRouteRequestTitle" type="hidden" value="Alcona County clerk"/>
<input id="hfClerkAddress" name="hfClerkAddress" type="hidden" value="106 5th Street"/>
<input id="hfClerkCityStateZip" name="hfClerkCityStateZip" type="hidden" value="Harrisville, Michigan  48740-0308"/>
<input id="hfCountyID" name="hfCountyID" type="hidden"

## Then test township look-up

In [7]:
payload = {'jurisdictionCode':'01040'}
res = requests.post('https://mvic.sos.state.mi.us/Voter/LocalClerk', data=payload, verify=False)

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,


In [8]:
res.status_code

200

In [9]:
type(res)

requests.models.Response

In [10]:
soup = BeautifulSoup(res.content, 'html.parser')

In [11]:
soup


<input id="hfPageTitle" name="hfPageTitle" type="hidden" value="Your city/township clerk"/><script async="" defer="" src="https://www.bing.com/api/maps/mapcontrol?callback=GetMap" type="text/javascript"></script>
<script src="/bundles/BingMapJS?v=XgqZHpFMh6PgxqFv2EbFUE-JrwdZK88VAIk5Ee20hLA1"></script>
<input id="hfPageTitle" name="hfPageTitle" type="hidden" value="Your city/township clerk"/>
<input id="hfPageName" name="hfPageName" type="hidden" value="clerk"/>
<input id="hfClerkLatitude" name="hfClerkLatitude" type="hidden" value="44.800836"/>
<input id="hfClerkLongitude" name="hfClerkLongitude" type="hidden" value="-83.328997"/>
<input id="hfClerkRouteRequestTitle" name="hfClerkRouteRequestTitle" type="hidden" value="Alcona Township clerk"/>
<input id="hfClerkAddress" name="hfClerkAddress" type="hidden" value="4892 Lavergne Rd"/>
<input id="hfClerkCityStateZip" name="hfClerkCityStateZip" type="hidden" value="Black River, Michigan  48721"/>
<input id="hfCountyID" name="hfCountyID" ty

# Scrape all county pages

In [12]:
tot_1 = str()

In [13]:
for i in range(1,84):
    response = requests.post(f"https://mvic.sos.state.mi.us/Voter/SearchByCounty?CountyID={i}", verify=False)
    tot_1 += response.text

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib

/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning,
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/urllib3/connectionpool.py:986: InsecureRequestWarning: Unverified HTTPS request is being made to host 'mvic.sos.state.mi.us'. Adding certificate verification is strongly advised. See: https://urllib

In [14]:
len(tot_1)

759742

In [15]:
soup = BeautifulSoup(tot_1, 'html.parser')

# Analyze the county pages

In [16]:
len(soup.find_all(class_ = ["ccd-page-heading"]))

332

In [17]:
soup.find_all(class_ = ["ccd-page-heading"])[:20]

[<h1 class="ccd-page-heading">Your county clerk</h1>,
 <h2 class="ccd-page-heading" id="lblLocalClerkLinksHeader">Alcona County city/township clerks </h2>,
 <h2 class="ccd-page-heading">
 Alcona County					</h2>,
 <h2 class="ccd-page-heading">
 Alcona County clerk's office
 					</h2>,
 <h1 class="ccd-page-heading">Your county clerk</h1>,
 <h2 class="ccd-page-heading" id="lblLocalClerkLinksHeader">Alger County city/township clerks </h2>,
 <h2 class="ccd-page-heading">
 Alger County					</h2>,
 <h2 class="ccd-page-heading">
 Alger County clerk's office
 					</h2>,
 <h1 class="ccd-page-heading">Your county clerk</h1>,
 <h2 class="ccd-page-heading" id="lblLocalClerkLinksHeader">Allegan County city/township clerks </h2>,
 <h2 class="ccd-page-heading">
 Allegan County					</h2>,
 <h2 class="ccd-page-heading">
 Allegan County clerk's office
 					</h2>,
 <h1 class="ccd-page-heading">Your county clerk</h1>,
 <h2 class="ccd-page-heading" id="lblLocalClerkLinksHeader">Alpena County city/townsh

In [18]:
len(soup.find_all(class_ = ["card-body"]))

249

In [19]:
soup.find_all(class_ = ["card-body"])[9:13]

[<div class="card-body">
 <div class="row"><div class="col-12 col-sm-6 col-md-4 col-lg-3 col-xl-2"><a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=01740&amp;countyID=4&amp;jurisdictionName=Alpena%20City">Alpena City</a></div><div class="col-12 col-sm-6 col-md-4 col-lg-3 col-xl-2"><a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=01760&amp;countyID=4&amp;jurisdictionName=Alpena%20Township">Alpena Township</a></div><div class="col-12 col-sm-6 col-md-4 col-lg-3 col-xl-2"><a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=34740&amp;countyID=4&amp;jurisdictionName=Green%20Township">Green Township</a></div><div class="col-12 col-sm-6 col-md-4 col-lg-3 col-xl-2"><a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=49360&amp;countyID=4&amp;jurisdictionName=Long%20Rapids%20Township">Long Rapids Township</a></div><div class="col-12 col-sm-6 col-md-4 col-lg-3 col-xl-2"><a clas

It appears that there are 3 "card-body" classes on each county page:
- a list of townships in that county
- contact details for the county clerk
- a map

# Build a list of all county names

In [20]:
soup.find_all(id_="lblLocalClerkLinksHeader")

[]

I wonder why soup doesn't find these id's?

In [21]:
divs = soup.find_all(class_ = ["ccd-page-heading"])

In [22]:
[(num,row) for num,row in enumerate(divs[:20])]

[(0, <h1 class="ccd-page-heading">Your county clerk</h1>),
 (1,
  <h2 class="ccd-page-heading" id="lblLocalClerkLinksHeader">Alcona County city/township clerks </h2>),
 (2,
  <h2 class="ccd-page-heading">
  Alcona County					</h2>),
 (3,
  <h2 class="ccd-page-heading">
  Alcona County clerk's office
  					</h2>),
 (4, <h1 class="ccd-page-heading">Your county clerk</h1>),
 (5,
  <h2 class="ccd-page-heading" id="lblLocalClerkLinksHeader">Alger County city/township clerks </h2>),
 (6,
  <h2 class="ccd-page-heading">
  Alger County					</h2>),
 (7,
  <h2 class="ccd-page-heading">
  Alger County clerk's office
  					</h2>),
 (8, <h1 class="ccd-page-heading">Your county clerk</h1>),
 (9,
  <h2 class="ccd-page-heading" id="lblLocalClerkLinksHeader">Allegan County city/township clerks </h2>),
 (10,
  <h2 class="ccd-page-heading">
  Allegan County					</h2>),
 (11,
  <h2 class="ccd-page-heading">
  Allegan County clerk's office
  					</h2>),
 (12, <h1 class="ccd-page-heading">Your county cle

In [23]:
divs[0].text

'Your county clerk'

In [24]:
counties = []
for county in range(int(len(divs)/4)):
    county_name = divs[2+4*county].text
    county_name = re.findall('\n(.*?)\t' ,county_name)[0]
    counties.append(county_name)

In [25]:
len(counties)

83

In [26]:
counties[-10:]

['St. Clair County',
 'St. Joseph County',
 'Sanilac County',
 'Schoolcraft County',
 'Shiawassee County',
 'Tuscola County',
 'Van Buren County',
 'Washtenaw County',
 'Wayne County',
 'Wexford County']

# Build a list of all jurisdictionCodes

In [27]:
divs = soup.find_all(class_ = "card-body")
# divs = soup.find_all(class_ = "local-clerk-link")

In [28]:
len(divs)

249

In [29]:
divs[0]

<div class="card-body">
<div class="row"><div class="col-12 col-sm-6 col-md-4 col-lg-3 col-xl-2"><a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=01040&amp;countyID=1&amp;jurisdictionName=Alcona%20Township">Alcona Township</a></div><div class="col-12 col-sm-6 col-md-4 col-lg-3 col-xl-2"><a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=12460&amp;countyID=1&amp;jurisdictionName=Caledonia%20Township">Caledonia Township</a></div><div class="col-12 col-sm-6 col-md-4 col-lg-3 col-xl-2"><a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=19320&amp;countyID=1&amp;jurisdictionName=Curtis%20Township">Curtis Township</a></div><div class="col-12 col-sm-6 col-md-4 col-lg-3 col-xl-2"><a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=34820&amp;countyID=1&amp;jurisdictionName=Greenbush%20Township">Greenbush Township</a></div><div class="col-12 col-sm-6 col-md-4 col-lg-3 col-xl-2"

In [30]:
divs[0].find_all(class_ = 'local-clerk-link')

[<a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=01040&amp;countyID=1&amp;jurisdictionName=Alcona%20Township">Alcona Township</a>,
 <a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=12460&amp;countyID=1&amp;jurisdictionName=Caledonia%20Township">Caledonia Township</a>,
 <a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=19320&amp;countyID=1&amp;jurisdictionName=Curtis%20Township">Curtis Township</a>,
 <a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=34820&amp;countyID=1&amp;jurisdictionName=Greenbush%20Township">Greenbush Township</a>,
 <a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=35740&amp;countyID=1&amp;jurisdictionName=Gustin%20Township">Gustin Township</a>,
 <a class="local-clerk-link" href="/Clerk/LocalClerk?dummy=x&amp;jurisdictionCode=36860&amp;countyID=1&amp;jurisdictionName=Harrisville%20City">Harrisville City</a>,
 <

In [31]:
juriCodes = []
for div in divs:
    if re.search('jurisdictionCode=(.*?)&amp', str(div)):
        juriCode = re.findall('jurisdictionCode=(.*?)&amp', str(div))
        juriCodes.append(juriCode)

In [32]:
len(juriCodes)

83

In [33]:
juriCodes

[['01040',
  '12460',
  '19320',
  '34820',
  '35740',
  '36860',
  '36880',
  '37220',
  '37380',
  '53880',
  '54060',
  '54740'],
 ['04460',
  '11960',
  '33380',
  '47540',
  '52360',
  '56200',
  '56220',
  '60820',
  '69160'],
 ['01260',
  '01280',
  '13700',
  '15200',
  '16720',
  '22680',
  '27740',
  '28120',
  '31360',
  '35720',
  '37460',
  '38640',
  '39200',
  '45180',
  '46600',
  '46760',
  '50840',
  '52000',
  '55200',
  '61620',
  '61640',
  '61820',
  '64740',
  '71100',
  '71700',
  '71720',
  '74980',
  '80620',
  '81580',
  '22740',
  '84580',
  '84880',
  '84900'],
 ['01740',
  '01760',
  '34740',
  '49360',
  '51220',
  '61520',
  '71240',
  '85180',
  '87680'],
 ['05240',
  '14420',
  '15380',
  '19400',
  '24640',
  '25340',
  '29600',
  '37500',
  '41960',
  '42400',
  '50640',
  '54440',
  '76260',
  '80100',
  '83300'],
 ['00280',
  '03360',
  '04120',
  '04140',
  '16240',
  '21120',
  '47580',
  '52120',
  '54800',
  '60660',
  '74020',
  '76120',
  '76

# Build a list of all township names

In [34]:
townships = []
for div in divs:
    if re.search('jurisdictionCode=(.*?)&amp', str(div)):
        township = re.findall('\">(.*?)</a', str(div.find_all(class_='local-clerk-link')))
        townships.append(township)

In [35]:
townships

[['Alcona Township',
  'Caledonia Township',
  'Curtis Township',
  'Greenbush Township',
  'Gustin Township',
  'Harrisville City',
  'Harrisville Township',
  'Hawes Township',
  'Haynes Township',
  'Mikado Township',
  'Millen Township',
  'Mitchell Township'],
 ['Au Train Township',
  'Burt Township',
  'Grand Island Township',
  'Limestone Township',
  'Mathias Township',
  'Munising City',
  'Munising Township',
  'Onota Township',
  'Rock River Township'],
 ['Allegan City',
  'Allegan Township',
  'Casco Township',
  'Cheshire Township',
  'Clyde Township',
  'Dorr Township',
  'Fennville City',
  'Fillmore Township',
  'Ganges Township',
  'Gunplain Township',
  'Heath Township',
  'Holland City',
  'Hopkins Township',
  'Laketown Township',
  'Lee Township',
  'Leighton Township',
  'Manlius Township',
  'Martin Township',
  'Monterey Township',
  'Otsego City',
  'Otsego Township',
  'Overisel Township',
  'Plainwell City',
  'Salem Township',
  'Saugatuck City',
  'Saugatuc

In [36]:
len(townships)

83

# Scrape all township (jurisdictions) pages

In [38]:
import warnings
warnings.simplefilter('ignore')

tot_2 = str()
for juriCode_list in juriCodes:
    for juriCode in juriCode_list:
        response = requests.post(f"https://mvic.sos.state.mi.us/Voter/LocalClerk?jurisdictionCode={juriCode}", verify=False)
        tot_2 += response.text

In [39]:
print("{:,.0f}".format(len(tot_2)))

8,171,034


In [40]:
# with open("Michigan_town.html", "w") as file:
#     file.write(tot_2)

In [41]:
township_data = BeautifulSoup(tot_2)

# Analyze the township pages

In [42]:
len(township_data)

1

In [43]:
type(township_data)

bs4.BeautifulSoup

In [44]:
township_cards = township_data.find_all(class_='card-body')

In [45]:
len(township_cards)

3070

There are 3070 containers with class="card-body" over 1535 township pages, 2 per township. One is for text data, and the other is for a map.

In [46]:
township_data.find(class_='additional-location-badge').find_next_siblings()

[<br/>,
 <br/>,
 <br/>,
 <br/>,
 <strong>Hours:</strong>,
 <div class="clerk-hours">Available 24 Hours/7 Days a Week<br/></div>,
 <br/>,
 <br/>,
 <a href="/Clerk/SearchByCounty?counties=1" id="county-clerk-link" style="font-size:1.2rem">County clerk information</a>]

## Isolate the textual data from the maps

In [47]:
township_dets = []
for x in range(0,len(township_cards), 2):
    township_dets.append(township_cards[x])

In [48]:
len(township_dets)

1535

In [49]:
township_dets[0]

<div class="card-body">
					Dawn Lalonde, Township Clerk
					<br/>
<br/>
					4892 Lavergne Rd
					<br/>
					Black River, Michigan  48721

					<br/>
<br/>
<strong>Phone:  </strong>(989) 471-2948
					<br/>
<strong>Fax:  </strong>(989) 471-2016
					<br/>
<strong>Email:  </strong>
<a href="mailto:alconaclerk@yahoo.com">alconaclerk@yahoo.com</a>
<br/>
<strong>11/3/2020<br/></strong> <strong>Pre-election weekend hours:</strong><div class="clerk-hours"><span style="font-weight:600">Saturday, October 31, 2020</span><br/> 7:30am-3:30pm<br/></div> <br/>
<span class="badge bg-gradient additional-location-badge">Drop box locations</span>
<br/>
<br/>5576 Us 23<br/>Black River, Michigan 48721<br/><strong>Hours:</strong><div class="clerk-hours">Available 24 Hours/7 Days a Week<br/></div>
<br/><br/><a href="/Clerk/SearchByCounty?counties=1" id="county-clerk-link" style="font-size:1.2rem">County clerk information</a>
</div>

In [50]:
township_dets[0].text

'\r\n\t\t\t\t\tDawn Lalonde, Township Clerk\r\n\t\t\t\t\t\n\r\n\t\t\t\t\t4892 Lavergne Rd\r\n\t\t\t\t\t\r\n\t\t\t\t\tBlack River, Michigan  48721\r\n\r\n\t\t\t\t\t\n\nPhone:\xa0\xa0(989) 471-2948\r\n\t\t\t\t\t\nFax:\xa0\xa0(989) 471-2016\r\n\t\t\t\t\t\nEmail:\xa0\xa0\nalconaclerk@yahoo.com\n\n11/3/2020 Pre-election weekend hours:Saturday, October 31, 2020 7:30am-3:30pm \nDrop box locations\n\n5576 Us 23Black River, Michigan\xa048721Hours:Available 24 Hours/7 Days a Week\nCounty clerk information\n'

In [51]:
type(township_dets[0])

bs4.element.Tag

In [52]:
[x.find_previous() for x in township_dets[0].find_all('br')]

[<div class="card-body">
 					Dawn Lalonde, Township Clerk
 					<br/>
 <br/>
 					4892 Lavergne Rd
 					<br/>
 					Black River, Michigan  48721
 
 					<br/>
 <br/>
 <strong>Phone:  </strong>(989) 471-2948
 					<br/>
 <strong>Fax:  </strong>(989) 471-2016
 					<br/>
 <strong>Email:  </strong>
 <a href="mailto:alconaclerk@yahoo.com">alconaclerk@yahoo.com</a>
 <br/>
 <strong>11/3/2020<br/></strong> <strong>Pre-election weekend hours:</strong><div class="clerk-hours"><span style="font-weight:600">Saturday, October 31, 2020</span><br/> 7:30am-3:30pm<br/></div> <br/>
 <span class="badge bg-gradient additional-location-badge">Drop box locations</span>
 <br/>
 <br/>5576 Us 23<br/>Black River, Michigan 48721<br/><strong>Hours:</strong><div class="clerk-hours">Available 24 Hours/7 Days a Week<br/></div>
 <br/><br/><a href="/Clerk/SearchByCounty?counties=1" id="county-clerk-link" style="font-size:1.2rem">County clerk information</a>
 </div>,
 <br/>,
 <br/>,
 <br/>,
 <br/>,
 <strong>Phone:

In [53]:
str(township_dets[0])

'<div class="card-body">\r\n\t\t\t\t\tDawn Lalonde, Township Clerk\r\n\t\t\t\t\t<br/>\n<br/>\r\n\t\t\t\t\t4892 Lavergne Rd\r\n\t\t\t\t\t<br/>\r\n\t\t\t\t\tBlack River, Michigan  48721\r\n\r\n\t\t\t\t\t<br/>\n<br/>\n<strong>Phone:\xa0\xa0</strong>(989) 471-2948\r\n\t\t\t\t\t<br/>\n<strong>Fax:\xa0\xa0</strong>(989) 471-2016\r\n\t\t\t\t\t<br/>\n<strong>Email:\xa0\xa0</strong>\n<a href="mailto:alconaclerk@yahoo.com">alconaclerk@yahoo.com</a>\n<br/>\n<strong>11/3/2020<br/></strong> <strong>Pre-election weekend hours:</strong><div class="clerk-hours"><span style="font-weight:600">Saturday, October 31, 2020</span><br/> 7:30am-3:30pm<br/></div> <br/>\n<span class="badge bg-gradient additional-location-badge">Drop box locations</span>\n<br/>\n<br/>5576 Us 23<br/>Black River, Michigan\xa048721<br/><strong>Hours:</strong><div class="clerk-hours">Available 24 Hours/7 Days a Week<br/></div>\n<br/><br/><a href="/Clerk/SearchByCounty?counties=1" id="county-clerk-link" style="font-size:1.2rem">County

In [54]:
print(township_dets[0])

<div class="card-body">
					Dawn Lalonde, Township Clerk
					<br/>
<br/>
					4892 Lavergne Rd
					<br/>
					Black River, Michigan  48721

					<br/>
<br/>
<strong>Phone:  </strong>(989) 471-2948
					<br/>
<strong>Fax:  </strong>(989) 471-2016
					<br/>
<strong>Email:  </strong>
<a href="mailto:alconaclerk@yahoo.com">alconaclerk@yahoo.com</a>
<br/>
<strong>11/3/2020<br/></strong> <strong>Pre-election weekend hours:</strong><div class="clerk-hours"><span style="font-weight:600">Saturday, October 31, 2020</span><br/> 7:30am-3:30pm<br/></div> <br/>
<span class="badge bg-gradient additional-location-badge">Drop box locations</span>
<br/>
<br/>5576 Us 23<br/>Black River, Michigan 48721<br/><strong>Hours:</strong><div class="clerk-hours">Available 24 Hours/7 Days a Week<br/></div>
<br/><br/><a href="/Clerk/SearchByCounty?counties=1" id="county-clerk-link" style="font-size:1.2rem">County clerk information</a>
</div>


In [55]:
str(township_dets[0]).splitlines()

['<div class="card-body">',
 '\t\t\t\t\tDawn Lalonde, Township Clerk',
 '\t\t\t\t\t<br/>',
 '<br/>',
 '\t\t\t\t\t4892 Lavergne Rd',
 '\t\t\t\t\t<br/>',
 '\t\t\t\t\tBlack River, Michigan  48721',
 '',
 '\t\t\t\t\t<br/>',
 '<br/>',
 '<strong>Phone:\xa0\xa0</strong>(989) 471-2948',
 '\t\t\t\t\t<br/>',
 '<strong>Fax:\xa0\xa0</strong>(989) 471-2016',
 '\t\t\t\t\t<br/>',
 '<strong>Email:\xa0\xa0</strong>',
 '<a href="mailto:alconaclerk@yahoo.com">alconaclerk@yahoo.com</a>',
 '<br/>',
 '<strong>11/3/2020<br/></strong> <strong>Pre-election weekend hours:</strong><div class="clerk-hours"><span style="font-weight:600">Saturday, October 31, 2020</span><br/> 7:30am-3:30pm<br/></div> <br/>',
 '<span class="badge bg-gradient additional-location-badge">Drop box locations</span>',
 '<br/>',
 '<br/>5576 Us 23<br/>Black River, Michigan\xa048721<br/><strong>Hours:</strong><div class="clerk-hours">Available 24 Hours/7 Days a Week<br/></div>',
 '<br/><br/><a href="/Clerk/SearchByCounty?counties=1" id="coun

In [56]:
township_dets[0].text.split('\r\n\t\t\t\t')

['',
 '\tDawn Lalonde, Township Clerk',
 '\t\n',
 '\t4892 Lavergne Rd',
 '\t',
 '\tBlack River, Michigan  48721\r\n',
 '\t\n\nPhone:\xa0\xa0(989) 471-2948',
 '\t\nFax:\xa0\xa0(989) 471-2016',
 '\t\nEmail:\xa0\xa0\nalconaclerk@yahoo.com\n\n11/3/2020 Pre-election weekend hours:Saturday, October 31, 2020 7:30am-3:30pm \nDrop box locations\n\n5576 Us 23Black River, Michigan\xa048721Hours:Available 24 Hours/7 Days a Week\nCounty clerk information\n']

In [57]:
township_data = [township_det.text.split('\r\n\t\t\t\t') for township_det in township_dets]

In [58]:
len(township_data)

1535

# Disagregate address elements into separate records

In [59]:
township_data_df = pd.DataFrame(township_data)

In [60]:
township_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1535 entries, 0 to 1534
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       1535 non-null   object
 1   1       1535 non-null   object
 2   2       1535 non-null   object
 3   3       1535 non-null   object
 4   4       1535 non-null   object
 5   5       1535 non-null   object
 6   6       1535 non-null   object
 7   7       1535 non-null   object
 8   8       1044 non-null   object
dtypes: object(9)
memory usage: 108.1+ KB


In [61]:
township_data_df.head()

0                                 1     2                      3  \
0      \tDawn Lalonde, Township Clerk  \t\n     \t4892 Lavergne Rd   
1       \tKerry Scott, Township Clerk  \t\n     \t6252 Hansen Road   
2    \tTeresa Perkins, Township Clerk  \t\n            \t4875 F-30   
3     \tShannon Story, Township Clerk  \t\n  \t5070 Woodland Drive   
4    \tRenee Lavergne, Township Clerk  \t\n        \t113 S. Second   

                              4                                   5  \
0                            \t  \tBlack River, Michigan  48721\r\n   
1  \t\r\nP.o. Box 7\t\t\t\t\t\t       \tSpruce, Michigan  48762\r\n   
2   \t\r\nPO Box 90\t\t\t\t\t\t      \tGlennie, Michigan  48737\r\n   
3    \t\r\nPO Box 9\t\t\t\t\t\t    \tGreenbush, Michigan  48738\r\n   
4   \t\r\nPO Box 25\t\t\t\t\t\t      \tLincoln, Michigan  48742\r\n   

                                                   6  \
0                       \t\n\nPhone:  (989) 471-2948   
1  \t\t\n\nMailing address:\n\n6461 Gillard Rd.\n...   
2  \t\t\n\nMailing address:\n\nPO Box 90\n\nGlenn...   
3  \t\t\n\nMailing address:\n\nP.o. Box 9\n\nGree...   
4  \t\t\n\nMailing address:\n\n935 N. Somers Rd.\...   

                                                   7  \
0                           \t\nFax:  (989) 471-2016   
1  \t\nFax:  \n\nEmail:  \nkerryscottclerk@gmail....   
2  \t\nFax:  \n\nEmail:  \ncurtistownshipclerk@ho...   
3  \t\nFax:  \n\nEmail:  \nshannonstory@charter.n...   
4  \t\nFax:  \n\nEmail:  \ngustintownshipclerk@gm...   

                                                   8  
0  \t\nEmail:  \nalconaclerk@yahoo.com\n\n11/3/20...  
1                                               None  
2                                               None  
3                                               None  
4                                               None

In [62]:
township_data_df[1].apply(lambda x: x.replace('\t', '').split(',')[0])

0            Dawn Lalonde
1             Kerry Scott
2          Teresa Perkins
3           Shannon Story
4          Renee Lavergne
              ...        
1530         Carol Perrin
1531    Tammy Porterfield
1532        Julie Cieslak
1533     Jessica Boonstra
1534       Teresa Hankins
Name: 1, Length: 1535, dtype: object

In [63]:
def fix_clerk(x):
    try:
        return x.replace('\t', '').split(',')[1].strip()
    except IndexError:
        return np.nan

In [64]:
township_data_df[1].apply(fix_clerk)

0       Township Clerk
1       Township Clerk
2       Township Clerk
3       Township Clerk
4       Township Clerk
             ...      
1530             Clerk
1531             Clerk
1532             Clerk
1533             Clerk
1534             Clerk
Name: 1, Length: 1535, dtype: object

In [65]:
township_data_df[3].apply(lambda x: x.replace('\t', ''))

0          4892 Lavergne Rd
1          6252 Hansen Road
2                 4875 F-30
3       5070 Woodland Drive
4             113 S. Second
               ...         
1530           3577 S 31 Rd
1531            4627 S 1 Rd
1532           7257 W 48 Rd
1533           4785 N. M-37
1534            6941 W 6 Rd
Name: 3, Length: 1535, dtype: object

In [66]:
township_data_df[4].apply(lambda x: re.findall('\n(.*?)\t', x))

0                 []
1       [P.o. Box 7]
2        [PO Box 90]
3         [PO Box 9]
4        [PO Box 25]
            ...     
1530              []
1531              []
1532              []
1533    [PO Box 323]
1534              []
Name: 4, Length: 1535, dtype: object

In [67]:
def fix_address_2(x):
    try:
        return re.findall('\n(.*?)\t', x)[0]
    except IndexError:
        return np.nan

In [68]:
township_data_df[4].apply(fix_address_2)

0              NaN
1       P.o. Box 7
2        PO Box 90
3         PO Box 9
4        PO Box 25
           ...    
1530           NaN
1531           NaN
1532           NaN
1533    PO Box 323
1534           NaN
Name: 4, Length: 1535, dtype: object

In [69]:
township_data_df[5].apply(lambda x: re.findall('\t(.*?)\r\n', x)[0].split(',')[0])

0       Black River
1            Spruce
2           Glennie
3         Greenbush
4           Lincoln
           ...     
1530       Cadillac
1531      Harrietta
1532       Wellston
1533         Mesick
1534         Mesick
Name: 5, Length: 1535, dtype: object

In [70]:
township_data_df[5].apply(lambda x: re.findall('\t(.*?)\r\n', x)[0].split(',')[1].split('  ')[0].strip())

0       Michigan
1       Michigan
2       Michigan
3       Michigan
4       Michigan
          ...   
1530    Michigan
1531    Michigan
1532    Michigan
1533    Michigan
1534    Michigan
Name: 5, Length: 1535, dtype: object

In [71]:
township_data_df[5].apply(lambda x: re.findall('\t(.*?)\r\n', x)[0].split(',')[1].split('  ')[1].strip())

0            48721
1            48762
2            48737
3            48738
4            48742
           ...    
1530         49601
1531         49638
1532         49689
1533         49668
1534    49668-9103
Name: 5, Length: 1535, dtype: object

In [72]:
township_data_df[6].apply(lambda x: re.findall('Phone:(.*)', x)[0].strip())

0       (989) 471-2948
1       (989) 471-5377
2       (989) 370-8855
3       (989) 569-3519
4       (989) 335-0635
             ...      
1530    (231) 920-8525
1531    (231) 389-0610
1532    (231) 862-3671
1533    (231) 885-1392
1534    (231) 269-3293
Name: 6, Length: 1535, dtype: object

In [73]:
township_data_df[6].apply(lambda x: re.search('Fax:(.*)', x)).value_counts

<bound method IndexOpsMixin.value_counts of 0       None
1       None
2       None
3       None
4       None
        ... 
1530    None
1531    None
1532    None
1533    None
1534    None
Name: 6, Length: 1535, dtype: object>

In [74]:
township_data_df[7].apply(lambda x: re.findall('Email:\xa0\xa0\n(.*)\n', x))

0                                      []
1             [kerryscottclerk@gmail.com]
2       [curtistownshipclerk@hotmail.com]
3              [shannonstory@charter.net]
4         [gustintownshipclerk@gmail.com]
                      ...                
1530                                   []
1531           [slagletwpclerk@gmail.com]
1532              [thecieslaks@yahoo.com]
1533                                   []
1534        [wexfordtwpclerk@acegroup.cc]
Name: 7, Length: 1535, dtype: object

In [75]:
def fix_email(x):
    try:
        return re.findall('Email:\xa0\xa0\n(.*)\n', x)[0]
    except IndexError:
        return np.nan

In [76]:
township_data_df[7].apply(fix_email)

0                                   NaN
1             kerryscottclerk@gmail.com
2       curtistownshipclerk@hotmail.com
3              shannonstory@charter.net
4         gustintownshipclerk@gmail.com
                     ...               
1530                                NaN
1531           slagletwpclerk@gmail.com
1532              thecieslaks@yahoo.com
1533                                NaN
1534        wexfordtwpclerk@acegroup.cc
Name: 7, Length: 1535, dtype: object

In [77]:
township_data_df[8]

0       \t\nEmail:  \nalconaclerk@yahoo.com\n\n11/3/20...
1                                                    None
2                                                    None
3                                                    None
4                                                    None
                              ...                        
1530    \t\nEmail:  \nyttpkh@yahoo.com\n\n\nDrop box l...
1531                                                 None
1532                                                 None
1533    \t\nEmail:  \nspringvilleclerk@goacentek.net\n...
1534                                                 None
Name: 8, Length: 1535, dtype: object

In [78]:
str(township_data_df[8])

'0       \\t\\nEmail:\xa0\xa0\\nalconaclerk@yahoo.com\\n\\n11/3/20...\n1                                                    None\n2                                                    None\n3                                                    None\n4                                                    None\n                              ...                        \n1530    \\t\\nEmail:\xa0\xa0\\nyttpkh@yahoo.com\\n\\n\\nDrop box l...\n1531                                                 None\n1532                                                 None\n1533    \\t\\nEmail:\xa0\xa0\\nspringvilleclerk@goacentek.net\\n...\n1534                                                 None\nName: 8, Length: 1535, dtype: object'

In [79]:
str(township_data_df[8][0])

'\t\nEmail:\xa0\xa0\nalconaclerk@yahoo.com\n\n11/3/2020 Pre-election weekend hours:Saturday, October 31, 2020 7:30am-3:30pm \nDrop box locations\n\n5576 Us 23Black River, Michigan\xa048721Hours:Available 24 Hours/7 Days a Week\nCounty clerk information\n'

In [80]:
township_data_df[8].apply(lambda x: re.findall('Email(.*?)', str(x)))

0       []
1       []
2       []
3       []
4       []
        ..
1530    []
1531    []
1532    []
1533    []
1534    []
Name: 8, Length: 1535, dtype: object

In [81]:
township_data_df[8].apply(lambda x: str(x).split('Email:  ')[0])

0       \t\nEmail:  \nalconaclerk@yahoo.com\n\n11/3/20...
1                                                    None
2                                                    None
3                                                    None
4                                                    None
                              ...                        
1530    \t\nEmail:  \nyttpkh@yahoo.com\n\n\nDrop box l...
1531                                                 None
1532                                                 None
1533    \t\nEmail:  \nspringvilleclerk@goacentek.net\n...
1534                                                 None
Name: 8, Length: 1535, dtype: object

In [82]:
township_data_df[8].apply(lambda x: re.findall('Email:\xa0\xa0\n(.*?)\n', str(x).split('Email:  ')[0]))

0                [alconaclerk@yahoo.com]
1                                     []
2                                     []
3                                     []
4                                     []
                      ...               
1530                  [yttpkh@yahoo.com]
1531                                  []
1532                                  []
1533    [springvilleclerk@goacentek.net]
1534                                  []
Name: 8, Length: 1535, dtype: object

# Create a DataFrame to hold township address data

In [83]:
county_town = []
for county_num,township in enumerate(townships):
    county_name = counties[county_num]
    for town in township:
        town_name = town
        data = {'county_name': county_name, 'town_name': town_name}
        county_town.append(data)

In [84]:
len(county_town)

1535

In [85]:
all_data_df = pd.DataFrame(county_town)

In [86]:
all_data_df

county_name              town_name
0      Alcona County        Alcona Township
1      Alcona County     Caledonia Township
2      Alcona County        Curtis Township
3      Alcona County     Greenbush Township
4      Alcona County        Gustin Township
...              ...                    ...
1530  Wexford County         Selma Township
1531  Wexford County        Slagle Township
1532  Wexford County  South Branch Township
1533  Wexford County   Springville Township
1534  Wexford County       Wexford Township

[1535 rows x 2 columns]

In [87]:
all_data_df['name'] = township_data_df[1].apply(lambda x: x.replace('\t', '').split(',')[0])
all_data_df['position'] = township_data_df[1].apply(fix_clerk)
all_data_df['address_1'] = township_data_df[3].apply(lambda x: x.replace('\t', ''))
all_data_df['address_2'] = township_data_df[4].apply(fix_address_2)
all_data_df['city'] = township_data_df[5].apply(lambda x: re.findall('\t(.*?)\r\n', x)[0].split(',')[0])
all_data_df['state'] = 'MI'
all_data_df['zip'] = township_data_df[5].apply(lambda x: re.findall('\t(.*?)\r\n', x)[0].split(',')[1].split('  ')[1].strip())
all_data_df['phone'] = township_data_df[6].apply(lambda x: re.findall('Phone:(.*)', x)[0].strip())
all_data_df['email'] = township_data_df[7].apply(fix_email)
all_data_df['email_2'] = township_data_df[8].apply(lambda x: re.findall('Email:\xa0\xa0\n(.*?)\n', str(x).split('Email:  ')[0]))

In [88]:
all_data_df

county_name              town_name               name  \
0      Alcona County        Alcona Township       Dawn Lalonde   
1      Alcona County     Caledonia Township        Kerry Scott   
2      Alcona County        Curtis Township     Teresa Perkins   
3      Alcona County     Greenbush Township      Shannon Story   
4      Alcona County        Gustin Township     Renee Lavergne   
...              ...                    ...                ...   
1530  Wexford County         Selma Township       Carol Perrin   
1531  Wexford County        Slagle Township  Tammy Porterfield   
1532  Wexford County  South Branch Township      Julie Cieslak   
1533  Wexford County   Springville Township   Jessica Boonstra   
1534  Wexford County       Wexford Township     Teresa Hankins   

            position            address_1   address_2         city state  \
0     Township Clerk     4892 Lavergne Rd         NaN  Black River    MI   
1     Township Clerk     6252 Hansen Road  P.o. Box 7       Spruce    MI   
2     Township Clerk            4875 F-30   PO Box 90      Glennie    MI   
3     Township Clerk  5070 Woodland Drive    PO Box 9    Greenbush    MI   
4     Township Clerk        113 S. Second   PO Box 25      Lincoln    MI   
...              ...                  ...         ...          ...   ...   
1530           Clerk         3577 S 31 Rd         NaN     Cadillac    MI   
1531           Clerk          4627 S 1 Rd         NaN    Harrietta    MI   
1532           Clerk         7257 W 48 Rd         NaN     Wellston    MI   
1533           Clerk         4785 N. M-37  PO Box 323       Mesick    MI   
1534           Clerk          6941 W 6 Rd         NaN       Mesick    MI   

             zip           phone                            email  \
0          48721  (989) 471-2948                              NaN   
1          48762  (989) 471-5377        kerryscottclerk@gmail.com   
2          48737  (989) 370-8855  curtistownshipclerk@hotmail.com   
3          48738  (989) 569-3519         shannonstory@charter.net   
4          48742  (989) 335-0635    gustintownshipclerk@gmail.com   
...          ...             ...                              ...   
1530       49601  (231) 920-8525                              NaN   
1531       49638  (231) 389-0610         slagletwpclerk@gmail.com   
1532       49689  (231) 862-3671            thecieslaks@yahoo.com   
1533       49668  (231) 885-1392                              NaN   
1534  49668-9103  (231) 269-3293      wexfordtwpclerk@acegroup.cc   

                               email_2  
0              [alconaclerk@yahoo.com]  
1                                   []  
2                                   []  
3                                   []  
4                                   []  
...                                ...  
1530                [yttpkh@yahoo.com]  
1531                                []  
1532                                []  
1533  [springvilleclerk@goacentek.net]  
1534                                []  

[1535 rows x 12 columns]

In [89]:
def fix_email_2(x):
    try:
        return x[0]
    except IndexError:
        return np.nan

In [90]:
all_data_df['email_2'] = all_data_df['email_2'].apply(fix_email_2)

In [91]:
for x in range(len(all_data_df['email'])):
    if pd.isnull(all_data_df['email'][x]):
        all_data_df['email'][x]=all_data_df['email_2'][x]

In [92]:
all_data_df = all_data_df.drop('email_2', axis=1)

In [93]:
all_data_df

county_name              town_name               name  \
0      Alcona County        Alcona Township       Dawn Lalonde   
1      Alcona County     Caledonia Township        Kerry Scott   
2      Alcona County        Curtis Township     Teresa Perkins   
3      Alcona County     Greenbush Township      Shannon Story   
4      Alcona County        Gustin Township     Renee Lavergne   
...              ...                    ...                ...   
1530  Wexford County         Selma Township       Carol Perrin   
1531  Wexford County        Slagle Township  Tammy Porterfield   
1532  Wexford County  South Branch Township      Julie Cieslak   
1533  Wexford County   Springville Township   Jessica Boonstra   
1534  Wexford County       Wexford Township     Teresa Hankins   

            position            address_1   address_2         city state  \
0     Township Clerk     4892 Lavergne Rd         NaN  Black River    MI   
1     Township Clerk     6252 Hansen Road  P.o. Box 7       Spruce    MI   
2     Township Clerk            4875 F-30   PO Box 90      Glennie    MI   
3     Township Clerk  5070 Woodland Drive    PO Box 9    Greenbush    MI   
4     Township Clerk        113 S. Second   PO Box 25      Lincoln    MI   
...              ...                  ...         ...          ...   ...   
1530           Clerk         3577 S 31 Rd         NaN     Cadillac    MI   
1531           Clerk          4627 S 1 Rd         NaN    Harrietta    MI   
1532           Clerk         7257 W 48 Rd         NaN     Wellston    MI   
1533           Clerk         4785 N. M-37  PO Box 323       Mesick    MI   
1534           Clerk          6941 W 6 Rd         NaN       Mesick    MI   

             zip           phone                            email  
0          48721  (989) 471-2948            alconaclerk@yahoo.com  
1          48762  (989) 471-5377        kerryscottclerk@gmail.com  
2          48737  (989) 370-8855  curtistownshipclerk@hotmail.com  
3          48738  (989) 569-3519         shannonstory@charter.net  
4          48742  (989) 335-0635    gustintownshipclerk@gmail.com  
...          ...             ...                              ...  
1530       49601  (231) 920-8525                 yttpkh@yahoo.com  
1531       49638  (231) 389-0610         slagletwpclerk@gmail.com  
1532       49689  (231) 862-3671            thecieslaks@yahoo.com  
1533       49668  (231) 885-1392   springvilleclerk@goacentek.net  
1534  49668-9103  (231) 269-3293      wexfordtwpclerk@acegroup.cc  

[1535 rows x 11 columns]

# Write township data to csv file

In [94]:
all_data_df.to_csv("Michigan_towns.csv", index=False)

# Disagregate dropbox address data

In [95]:
drop_box_data = township_data_df[8].apply(lambda x: re.findall('\n(.*?)\n', str(x).split('Email:  ')[0]))

In [96]:
type(drop_box_data)

pandas.core.series.Series

In [97]:
drop_box_data[:20]

0     [Email:  , , Drop box locations, 5576 Us 23Bla...
1                                                    []
2                                                    []
3                                                    []
4                                                    []
5     [Email:  , , 11/3/2020 Pre-election weekend ho...
6                                          [Email:  , ]
7                                                    []
8                                                    []
9                                                    []
10                                                   []
11                                                   []
12    [Email:  , , 11/3/2020 Pre-election weekend ho...
13    [Email:  , , 11/3/2020 Pre-election weekend ho...
14                                                   []
15    [Email:  , , 11/3/2020 Pre-election weekend ho...
16                                                   []
17    [Email:  , , 11/3/2020 Pre-election weeken

In [98]:
township_data_df.loc[2,7]

'\t\nFax:\xa0\xa0\n\nEmail:\xa0\xa0\ncurtistownshipclerk@hotmail.com\n\nBusiness hours:Mon. By AppointmentTue. By AppointmentWed. By AppointmentThu. By AppointmentFri. By Appointment\n11/3/2020 Pre-election weekend hours:Saturday, October 31, 2020 8am-4pm \nDrop box locations\n\n4875 F30Glennie, Michigan\xa048737Hours:Available 24 Hours/7 Days a Week\nCounty clerk information\n'

In [99]:
township_data_df.loc[0,8]

'\t\nEmail:\xa0\xa0\nalconaclerk@yahoo.com\n\n11/3/2020 Pre-election weekend hours:Saturday, October 31, 2020 7:30am-3:30pm \nDrop box locations\n\n5576 Us 23Black River, Michigan\xa048721Hours:Available 24 Hours/7 Days a Week\nCounty clerk information\n'

In [100]:
township_data_df.loc[0,8].splitlines()

['\t',
 'Email:\xa0\xa0',
 'alconaclerk@yahoo.com',
 '',
 '11/3/2020 Pre-election weekend hours:Saturday, October 31, 2020 7:30am-3:30pm ',
 'Drop box locations',
 '',
 '5576 Us 23Black River, Michigan\xa048721Hours:Available 24 Hours/7 Days a Week',
 'County clerk information']

# Determine patterns of data in dropbox rows

In [101]:
list_col_8 = []
for num,row in enumerate(township_data_df[8]):
    for r,s in enumerate(str(row).splitlines()):
        if re.search('Drop box locations', s):
            print(num, r)
            list_col_8.append(r)

0 5
12 6
15 6
18 6
19 5
26 6
27 6
28 6
30 6
31 6
34 6
35 6
37 6
38 6
40 6
42 6
43 6
45 6
47 6
49 6
51 6
52 6
53 6
54 6
55 6
56 6
60 6
63 6
64 6
68 6
69 6
70 6
72 5
76 6
80 9
88 6
92 6
96 6
97 6
101 6
103 6
104 6
105 6
106 6
107 6
109 6
111 6
112 6
113 6
114 6
117 6
118 6
119 6
121 5
124 6
126 6
128 6
130 5
131 5
132 6
138 6
141 6
142 6
147 6
149 6
150 6
151 6
152 6
153 6
154 6
155 6
156 6
157 5
158 5
159 6
160 6
161 6
162 6
163 6
164 6
166 6
169 6
170 6
171 5
172 6
173 6
176 6
179 6
180 6
184 6
185 6
195 6
196 6
197 6
198 6
199 6
200 6
201 6
202 6
203 6
205 6
206 6
207 6
208 6
209 6
210 6
211 6
212 6
213 6
214 6
215 6
216 6
217 6
221 5
227 6
228 6
229 5
232 6
234 6
237 5
240 5
241 5
243 6
244 5
247 5
252 6
254 6
255 6
262 6
263 5
274 6
277 5
278 6
281 9
283 6
284 6
285 6
287 6
291 6
293 6
294 6
295 6
299 6
301 6
306 6
308 6
311 5
312 6
313 6
316 9
320 6
321 6
324 6
326 6
330 6
331 6
333 5
334 6
335 5
346 6
350 6
352 6
353 6
354 6
355 5
356 6
357 6
358 6
362 6
365 6
367 6
368 6
370 6
37

In [102]:
pd.Series(list_col_8).value_counts()

6    684
5     57
9     22
dtype: int64

In [103]:
list_col_7=[]
for num,row in enumerate(township_data_df[7]):
    for r,s in enumerate(str(row).splitlines()):
        if re.search('Drop box locations', s):
            print(num, r)
            list_col_7.append(r)

1 7
2 8
7 7
9 7
10 7
11 8
16 8
21 10
23 8
24 8
29 8
32 8
33 8
48 7
62 8
66 8
78 7
82 7
87 8
93 8
94 11
98 8
120 8
123 7
127 8
129 8
136 8
140 8
143 8
187 8
191 8
204 8
222 8
233 8
235 7
238 8
239 8
242 8
253 7
264 11
268 8
270 11
271 8
275 8
296 8
298 8
300 8
302 8
305 8
310 8
315 8
317 8
318 7
323 7
332 8
342 8
344 8
359 7
361 8
364 8
366 8
373 8
379 7
380 8
381 8
384 8
391 8
392 8
393 8
396 8
399 10
400 8
415 8
418 8
434 7
436 8
440 7
441 7
442 7
443 7
460 8
463 8
465 8
467 8
470 7
471 7
479 8
483 8
486 8
501 8
504 8
505 8
508 8
517 7
528 7
539 7
543 11
549 7
550 8
555 7
582 7
585 10
587 8
594 7
600 8
610 7
620 8
625 8
628 8
634 8
647 8
652 8
653 8
656 8
660 8
679 8
684 8
692 8
699 8
702 8
713 8
727 8
734 8
736 8
738 11
748 8
753 8
763 8
785 7
787 8
790 7
792 8
793 8
812 8
826 8
859 8
861 8
869 8
873 8
900 11
905 8
907 8
912 8
921 8
941 7
958 8
964 8
970 8
977 8
979 8
993 8
998 8
1004 8
1005 8
1008 7
1011 8
1016 8
1020 7
1024 8
1055 8
1062 8
1073 8
1076 8
1102 8
1147 8
1151 8
1158 8


In [104]:
pd.Series(list_col_7).value_counts()

8     165
7      39
11      8
10      4
dtype: int64

In [105]:
township_data_df.loc[0,8].splitlines()[7].split('Hours:')

['5576 Us 23Black River, Michigan\xa048721',
 'Available 24 Hours/7 Days a Week']

In [106]:
township_data_df.loc[15,8].splitlines()[8].split('Hours:')

['Limestone Township HallN 3010 State Highway 67Limestone, Michigan\xa049816',
 'Available 24 Hours/7 Days a WeekLimestone Township HallN 3010 State Highway 67Limestone, Michigan\xa049816',
 'Available 24 Hours/7 Days a Week']

In [107]:
township_data_df.loc[2,7].splitlines()

['\t',
 'Fax:\xa0\xa0',
 '',
 'Email:\xa0\xa0',
 'curtistownshipclerk@hotmail.com',
 '',
 'Business hours:Mon. By AppointmentTue. By AppointmentWed. By AppointmentThu. By AppointmentFri. By Appointment',
 '11/3/2020 Pre-election weekend hours:Saturday, October 31, 2020 8am-4pm ',
 'Drop box locations',
 '',
 '4875 F30Glennie, Michigan\xa048737Hours:Available 24 Hours/7 Days a Week',
 'County clerk information']

In [108]:
township_data_df.loc[2,7].splitlines()[10].split('Hours:')

['4875 F30Glennie, Michigan\xa048737', 'Available 24 Hours/7 Days a Week']

In [109]:
township_data_df.loc[7,7].splitlines()[9].split('Hours:')

['1341 N Fawn StLincoln, Michigan\xa048742',
 'Mon. 10am-9pmTue. 10am-9pmWed. 10am-9pmThu. 10am-9pmFri. 10am-9pmSat. 10am-9pmSun. 12pm-9pm']

## Analyze first set

In [110]:
township_data_df.loc[0,8].splitlines()[7]

'5576 Us 23Black River, Michigan\xa048721Hours:Available 24 Hours/7 Days a Week'

In [111]:
township_data_df.loc[0,8].splitlines()[7].split('Hours:')[0].split(', Michigan\xa0')

['5576 Us 23Black River', '48721']

In [112]:
township_data_df.loc[0,8].splitlines()[7].split('Hours:')[0].split(', Michigan\xa0')[0]

'5576 Us 23Black River'

In [113]:
township_data_df.loc[15,8].splitlines()[8].split('Hours:')[0].split(', Michigan\xa0')[0]

'Limestone Township HallN 3010 State Highway 67Limestone'

In [114]:
township_data_df.loc[0,8].splitlines()[7].split('Hours:')[0].split(', Michigan\xa0')[1]

'48721'

In [115]:
township_data_df.loc[15,8].splitlines()[8].split('Hours:')[0].split(', Michigan\xa0')[1]

'49816'

In [116]:
township_data_df.loc[0,8].splitlines()[7].split('Hours:')[1]

'Available 24 Hours/7 Days a Week'

## Analyze second set

In [117]:
township_data_df

0                                  1     2                      3  \
0          \tDawn Lalonde, Township Clerk  \t\n     \t4892 Lavergne Rd   
1           \tKerry Scott, Township Clerk  \t\n     \t6252 Hansen Road   
2        \tTeresa Perkins, Township Clerk  \t\n            \t4875 F-30   
3         \tShannon Story, Township Clerk  \t\n  \t5070 Woodland Drive   
4        \tRenee Lavergne, Township Clerk  \t\n        \t113 S. Second   
...  ..                               ...   ...                    ...   
1530                \tCarol Perrin, Clerk  \t\n         \t3577 S 31 Rd   
1531           \tTammy Porterfield, Clerk  \t\n          \t4627 S 1 Rd   
1532               \tJulie Cieslak, Clerk  \t\n         \t7257 W 48 Rd   
1533            \tJessica Boonstra, Clerk  \t\n         \t4785 N. M-37   
1534              \tTeresa Hankins, Clerk  \t\n          \t6941 W 6 Rd   

                                 4                                   5  \
0                               \t  \tBlack River, Michigan  48721\r\n   
1     \t\r\nP.o. Box 7\t\t\t\t\t\t       \tSpruce, Michigan  48762\r\n   
2      \t\r\nPO Box 90\t\t\t\t\t\t      \tGlennie, Michigan  48737\r\n   
3       \t\r\nPO Box 9\t\t\t\t\t\t    \tGreenbush, Michigan  48738\r\n   
4      \t\r\nPO Box 25\t\t\t\t\t\t      \tLincoln, Michigan  48742\r\n   
...                            ...                                 ...   
1530                            \t     \tCadillac, Michigan  49601\r\n   
1531                            \t    \tHarrietta, Michigan  49638\r\n   
1532                            \t     \tWellston, Michigan  49689\r\n   
1533  \t\r\nPO Box 323\t\t\t\t\t\t       \tMesick, Michigan  49668\r\n   
1534                            \t  \tMesick, Michigan  49668-9103\r\n   

                                                      6  \
0                          \t\n\nPhone:  (989) 471-2948   
1     \t\t\n\nMailing address:\n\n6461 Gillard Rd.\n...   
2     \t\t\n\nMailing address:\n\nPO Box 90\n\nGlenn...   
3     \t\t\n\nMailing address:\n\nP.o. Box 9\n\nGree...   
4     \t\t\n\nMailing address:\n\n935 N. Somers Rd.\...   
...                                                 ...   
1530                       \t\n\nPhone:  (231) 920-8525   
1531                       \t\n\nPhone:  (231) 389-0610   
1532                       \t\n\nPhone:  (231) 862-3671   
1533                       \t\n\nPhone:  (231) 885-1392   
1534                       \t\n\nPhone:  (231) 269-3293   

                                                      7  \
0                              \t\nFax:  (989) 471-2016   
1     \t\nFax:  \n\nEmail:  \nkerryscottclerk@gmail....   
2     \t\nFax:  \n\nEmail:  \ncurtistownshipclerk@ho...   
3     \t\nFax:  \n\nEmail:  \nshannonstory@charter.n...   
4     \t\nFax:  \n\nEmail:  \ngustintownshipclerk@gm...   
...                                                 ...   
1530                           \t\nFax:  (231) 775-5071   
1531  \t\nFax:  \n\nEmail:  \nslagletwpclerk@gmail.c...   
1532  \t\nFax:  \n\nEmail:  \nthecieslaks@yahoo.com\...   
1533                           \t\nFax:  (231) 885-1909   
1534  \t\nFax:  \n\nEmail:  \nwexfordtwpclerk@acegro...   

                                                      8  
0     \t\nEmail:  \nalconaclerk@yahoo.com\n\n11/3/20...  
1                                                  None  
2                                                  None  
3                                                  None  
4                                                  None  
...                                                 ...  
1530  \t\nEmail:  \nyttpkh@yahoo.com\n\n\nDrop box l...  
1531                                               None  
1532                                               None  
1533  \t\nEmail:  \nspringvilleclerk@goacentek.net\n...  
1534                                               None  

[1535 rows x 9 columns]

In [118]:
township_data_df.loc[2,7].splitlines()[10].split('Hours:')[0].split(', Michigan\xa0')

['4875 F30Glennie', '48737']

In [119]:
township_data_df.loc[2,7].splitlines()[10].split('Hours:')[0].split(', Michigan\xa0')[0]

'4875 F30Glennie'

In [120]:
township_data_df.loc[2,7].splitlines()[10].split('Hours:')[0].split(', Michigan\xa0')[1]

'48737'

In [121]:
township_data_df.loc[2,7].splitlines()[10].split('Hours:')[1]

'Available 24 Hours/7 Days a Week'

# Loop thru all cases to get dropbox data

In [122]:
bool(re.search('Drop box locations', township_data_df[8][0]))

True

In [123]:
bool(re.search('Drop box locations', township_data_df[7][0]))

False

In [124]:
dropBox = []
for i in range(len(township_data_df)):
    if bool(re.search('Drop box locations', str(township_data_df[7][i]))) or bool(re.search('Drop box locations', str(township_data_df[8][i]))):
        has_dropBox = 'Yes'
        
        
        if bool(re.search('Drop box locations', str(township_data_df[7][i]))):
            if bool(re.search('Michigan', township_data_df[7][i].splitlines()[10])):
                address_1 = township_data_df[7][i].splitlines()[10].split('Hours:')[0].split(', Michigan\xa0')[0]
                zip_code = township_data_df[7][i].splitlines()[10].split('Hours:')[0].split(', Michigan\xa0')[1]
                try:
                    hours = township_data_df[7][i].splitlines()[10].split('Hours:')[1]
                except IndexError:
                    hours = np.nan
            elif bool(re.search('Michigan', township_data_df[7][i].splitlines()[9])):
                address_1 = township_data_df[7][i].splitlines()[9].split('Hours:')[0].split(', Michigan\xa0')[0]
                zip_code = township_data_df[7][i].splitlines()[9].split('Hours:')[0].split(', Michigan\xa0')[1]
                try:
                    hours = township_data_df[7][i].splitlines()[9].split('Hours:')[1]
                except IndexError:
                    hours=np.nan
        
        
        elif bool(re.search('Drop box locations', str(township_data_df[8][i]))):
            if bool(re.search('Michigan', township_data_df[8][i].splitlines()[7])):
                address_1 = township_data_df[8][i].splitlines()[7].split('Hours:')[0].split(', Michigan\xa0')[0]
                zip_code = township_data_df[8][i].splitlines()[7].split('Hours:')[0].split(', Michigan\xa0')[1]
                try:
                    hours = township_data_df[8][i].splitlines()[7].split('Hours:')[1]
                except IndexError:
                    hours = np.nan
            elif bool(re.search('Michigan', township_data_df[8][i].splitlines()[8])):
                address_1 = township_data_df[8][i].splitlines()[8].split('Hours:')[0].split(', Michigan\xa0')[0]
                zip_code = township_data_df[8][i].splitlines()[8].split('Hours:')[0].split(', Michigan\xa0')[1]
                try:
                    hours = township_data_df[8][i].splitlines()[8].split('Hours:')[1]
                except IndexError:
                    hours = np.nan
            
    else:
        has_dropBox = 'No'
        address_1 = np.nan
        zip_code = np.nan
        hours = np.nan
    
    data = {'has_dropBox': has_dropBox, 'address_1':address_1, 'zip': zip_code, 'hours': hours}
    dropBox.append(data)

In [125]:
len(dropBox)

1535

In [126]:
dropBox[:10]

[{'has_dropBox': 'Yes',
  'address_1': '5576 Us 23Black River',
  'zip': '48721',
  'hours': 'Available 24 Hours/7 Days a Week'},
 {'has_dropBox': 'Yes',
  'address_1': '6461 Gillard RdSpruce',
  'zip': '48762',
  'hours': '24/7'},
 {'has_dropBox': 'Yes',
  'address_1': '4875 F30Glennie',
  'zip': '48737',
  'hours': 'Available 24 Hours/7 Days a Week'},
 {'has_dropBox': 'No', 'address_1': nan, 'zip': nan, 'hours': nan},
 {'has_dropBox': 'No', 'address_1': nan, 'zip': nan, 'hours': nan},
 {'has_dropBox': 'No', 'address_1': nan, 'zip': nan, 'hours': nan},
 {'has_dropBox': 'No', 'address_1': nan, 'zip': nan, 'hours': nan},
 {'has_dropBox': 'Yes',
  'address_1': '1341 N Fawn StLincoln',
  'zip': '48742',
  'hours': 'Mon. 10am-9pmTue. 10am-9pmWed. 10am-9pmThu. 10am-9pmFri. 10am-9pmSat. 10am-9pmSun. 12pm-9pm'},
 {'has_dropBox': 'No', 'address_1': nan, 'zip': nan, 'hours': nan},
 {'has_dropBox': 'Yes',
  'address_1': '2268 Hamilton RdBlack Box By Garage Door.Mikado',
  'zip': '48745',
  'hour

# Create DataFrame to hold drop box address data

In [127]:
dropBox_df = pd.DataFrame(dropBox)

In [128]:
dropBox_df['county_name']=all_data_df['county_name']
dropBox_df['town_name']= all_data_df['town_name']

In [129]:
dropBox_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1535 entries, 0 to 1534
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   has_dropBox  1535 non-null   object
 1   address_1    979 non-null    object
 2   zip          979 non-null    object
 3   hours        929 non-null    object
 4   county_name  1535 non-null   object
 5   town_name    1535 non-null   object
dtypes: object(6)
memory usage: 72.1+ KB


In [130]:
# dropBox_df=pd.read_csv('Michigan_dropBox.csv')

In [131]:
dropBox_df

has_dropBox                                          address_1  \
0            Yes                              5576 Us 23Black River   
1            Yes                              6461 Gillard RdSpruce   
2            Yes                                    4875 F30Glennie   
3             No                                                NaN   
4             No                                                NaN   
...          ...                                                ...   
1530         Yes  4101 S 35 RdI Can Be Contacted At 231-920-8525...   
1531         Yes              Tammy Porterfield4627 S 1 RdHarrietta   
1532          No                                                NaN   
1533          No                                                NaN   
1534          No                                                NaN   

                                                    zip  \
0                                                 48721   
1                                                 48762   
2                                                 48737   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
1530                                              49601   
1531  49638Business hours:Mon. By AppointmentTue. By...   
1532                                                NaN   
1533                                                NaN   
1534                                                NaN   

                                 hours     county_name              town_name  
0     Available 24 Hours/7 Days a Week   Alcona County        Alcona Township  
1                                 24/7   Alcona County     Caledonia Township  
2     Available 24 Hours/7 Days a Week   Alcona County        Curtis Township  
3                                  NaN   Alcona County     Greenbush Township  
4                                  NaN   Alcona County        Gustin Township  
...                                ...             ...                    ...  
1530                              24/7  Wexford County         Selma Township  
1531                               NaN  Wexford County        Slagle Township  
1532                               NaN  Wexford County  South Branch Township  
1533                               NaN  Wexford County   Springville Township  
1534                               NaN  Wexford County       Wexford Township  

[1535 rows x 6 columns]

In [132]:
dropBox_df=dropBox_df.rename(columns={'address_1':'address'})

# Create dropbox Address and City 

In [133]:
def fix_dropBox_street(address):
    try:
        if re.search("\B[0-9][A-Z]{1}", address):
            x = re.findall("\B[0-9][A-Z]{1}", address)
            y = re.split("\B[0-9][A-Z]{1}", address, 1)
        elif re.search("\B[a-z][A-Z]{1}", address):
            x = re.findall("\B[a-z][A-Z]{1}", address)
            y = re.split("\B[a-z][A-Z]{1}", address, 1)
            
        street = y[0]+x[0][0]
#         city = x[0][1]+y[1]
        
        return street
    except:
        pass

In [134]:
def fix_dropBox_city(address):
    try:
        if re.search("\B[0-9][A-Z]{1}", address):
            x = re.findall("\B[0-9][A-Z]{1}", address)
            y = re.split("\B[0-9][A-Z]{1}", address, 1)
        elif re.search("\B[a-z][A-Z]{1}", address):
            x = re.findall("\B[a-z][A-Z]{1}", address)
            y = re.split("\B[a-z][A-Z]{1}", address, 1)
            
#         street = y[0]+x[0][0]
        city = x[0][1]+y[1]
        
        return city
    except:
        pass

In [204]:
dropBox_df['address_1'] = dropBox_df['address'].apply(fix_dropBox_street)

In [136]:
dropBox_df['city'] = dropBox_df['address'].apply(fix_dropBox_city)

# Create dropbox location_type

In [153]:
def loc_type(address):
    try:
        x = re.findall("\B[a-z][0-9]{1}", address)
        y = re.split("\B[a-z][0-9]{1}", address, 1)
        location_type = y[0]+x[0][0]
#         notes = x[0][1]+y[1]
        
        return location_type
    
    except:
        pass

In [205]:
dropBox_df['address_1'].apply(loc_type)

0                    None
1                    None
2                    None
3                    None
4                    None
              ...        
1530                 None
1531    Tammy Porterfield
1532                 None
1533                 None
1534                 None
Name: address_1, Length: 1535, dtype: object

In [206]:
dropBox_df['location_type']=dropBox_df['address_1'].apply(loc_type)

In [162]:
def add_1(address):
    try:
        x = re.findall("\B[a-z][0-9]{1}", address)
        y = re.split("\B[a-z][0-9]{1}", address, 1)
        address_1 = x[0][1]+y[1]
        
        return address_1
        
    except:
        return address

In [207]:
dropBox_df['address_1']=dropBox_df['address_1'].apply(add_1)

# Create dropbox notes

In [183]:
def notes(address):
    try:
        x = re.findall("\.[A-Z]{1}", address)
        y = re.split("\.[A-Z]{1}", address, 1)
#         location_type = y[0]+x[0][0]
        notes = x[0][1]+y[1]
        
        return notes
    
    except:
        pass

In [208]:
dropBox_df['notes']=dropBox_df['address_1'].apply(notes)

In [197]:
def add_2(address):
    try:
        x = re.findall("\.[A-Z]{1}", address)
        y = re.split("\.[A-Z]{1}", address, 1)
        add_2 = y[0]+x[0][0]
        
        return add_2
    
    except:
        return address

In [209]:
dropBox_df['address_1']=dropBox_df['address_1'].apply(add_2)

# Write dropbox file to csv file

In [210]:
dropBox_df.to_csv('Michigan_dropBox.csv', index=False)